File: Packages/js-common/Source/Utils/DB/RatingProcessor.ts
import {emptyObj, IsNumber, Assert, CE, emptyArray_forLoading, emptyArray} from "js-vextensions";
import {CreateAccessor, NoID, PartialBy} from "mobx-graphlink";
import {GetRating, GetRatingAverage, GetRatings} from "../../DB/nodeRatings.js";
import {NodeRating, NodeRating_MaybePseudo} from "../../DB/nodeRatings/@NodeRating.js";
import {NodeRatingType} from "../../DB/nodeRatings/@NodeRatingType.js";
import {GetMainRatingType, GetNodeForm, GetRatingTypesForNode} from "../../DB/nodes/$node.js";
import {NodeL1, NodeL2} from "../../DB/nodes/@Node.js";
import {ArgumentType} from "../../DB/nodes/@NodeRevision.js";
import {ClaimForm} from "../../DB/nodeLinks/@NodeLink.js";

// sync:rs
export const GetArgumentImpactPseudoRating = CreateAccessor((argument: NodeL1, premises: NodeL1[], userID: string, useAverageForMissing = false): PartialBy<NodeRating, "id" | "accessPolicy">|n=>{
	if (CE(premises).Any(a=>a == null)) return null; // must still be loading
	if (premises.length == 0) return null;

	const premiseProbabilities = [] as number[];
	for (const premise of premises) {
		//const ratingType = GetRatingTypesForNode(premise)[0].type;
		const ratingType = NodeRatingType.truth;
		let ratingValue = GetRating(premise.id, ratingType, userID)?.value ?? null;
		// if user didn't rate this premise, just use the average rating
		if (ratingValue == null) {
			if (useAverageForMissing) {
				ratingValue = GetRatingAverage(premise.id, ratingType, undefined) || 0;
			} else {
				return null;
			}
		}

		const form = GetNodeForm(premise, argument);
		const probability = form == ClaimForm.negation ? 1 - (ratingValue / 100) : (ratingValue / 100);
		premiseProbabilities.push(probability);
	}

	let combinedTruthOfPremises;
	if (argument.argumentType == ArgumentType.all) {
		combinedTruthOfPremises = premiseProbabilities.reduce((total, current)=>total * current, 1);
	} else if (argument.argumentType == ArgumentType.anyTwo) {
		const strongest = CE(premiseProbabilities).Max(undefined, true);
		const secondStrongest = premiseProbabilities.length > 1 ? CE(CE(premiseProbabilities).Exclude({excludeEachOnlyOnce: true}, strongest)).Max(undefined, true) : 0;
		combinedTruthOfPremises = strongest * secondStrongest;
	} else {
		combinedTruthOfPremises = CE(premiseProbabilities).Max(undefined, true);
	}

	let relevance = GetRating(argument.id, NodeRatingType.relevance, userID)?.value ?? null;
	// if user didn't rate the relevance, just use the average rating
	if (relevance == null) {
		if (useAverageForMissing) {
			relevance = GetRatingAverage(argument.id, NodeRatingType.relevance) || 0;
		} else {
			return null;
		}
	}

	// let strengthForType = adjustment.Distance(50) / 50;
	const result = combinedTruthOfPremises * (relevance / 100);
	Assert(IsNumber(result), `Impact pseudo-rating is null. @combinedTruthOfPremises:${combinedTruthOfPremises} @relevance:${relevance}`);

	return {
		//_key: userID,
		//accessPolicy: null,
		node: argument.id,
		type: NodeRatingType.impact,
		creator: userID,
		createdAt: Date.now(),
		value: CE(result * 100).RoundTo(1),
	};
});
// export function GetArgumentStrengthEntries(nodeChildren: NodeL1[], users: string[]) {
/* export function GetArgumentStrengthPseudoRatings(nodeChildren: NodeL1[]): Rating[] {
	if (nodeChildren.Any(a=>a == null)) return []; // must still be loading
	let impactPremise = nodeChildren.First(a=>a.impactPremise != null);
	let premises = nodeChildren.Except(impactPremise);
	if (premises.length == 0) return [];

	let usersWhoRated = nodeChildren.SelectMany(child=>GetRatings(child._id, NodeL1.GetMainRatingTypes(child)[0]).map(a=>a.id)).Distinct();
	let result = usersWhoRated.map(userID=>GetArgumentStrengthPseudoRating(nodeChildren, userID));
	return result;
} */

export function RatingListAfterRemovesAndAdds(baseList: NodeRating[], ratingsToRemove?: string[], ratingsToAdd?: NodeRating[], ratingsToAdd_filter?: {nodeID: string, ratingType: NodeRatingType, userIDs?: string[]|n}) {
	let result = baseList.slice();
	if (ratingsToRemove) result = result.filter(a=>!ratingsToRemove.includes(a.id));
	if (ratingsToAdd) {
		const filter = ratingsToAdd_filter;
		Assert(filter != null, "If supplying ratingsToAdd, must also supply ratingsToAdd_filter.");
		result.push(...ratingsToAdd.filter(a=>a.node == filter.nodeID && a.type == filter.ratingType && (filter.userIDs == null || filter.userIDs.includes(a.creator))));
	}
	return result;
}

// sync:rs[loosely]
export const GetArgumentImpactPseudoRatings = CreateAccessor((
	argument: NodeL1,
	premises: NodeL1[],
	userIDs?: string[]|n,
	useAverageForMissing = false,
	ratingsBeingRemoved?: string[],
	ratingsBeingAdded?: NodeRating[],
): NodeRating_MaybePseudo[]=>{
	if (CE(premises).Any(a=>a == null)) return emptyArray_forLoading as any; // must still be loading
	if (premises.length == 0) return emptyArray as any;

	let argumentRelevanceRatings = GetRatings(argument.id, NodeRatingType.relevance, userIDs);
	argumentRelevanceRatings = RatingListAfterRemovesAndAdds(argumentRelevanceRatings, ratingsBeingRemoved, ratingsBeingAdded, {nodeID: argument.id, ratingType: NodeRatingType.relevance, userIDs});

	const usersWhoRatedArgAndPremises = new Set(argumentRelevanceRatings.map(a=>a.creator));
	for (const premise of premises) {
		let premiseTruthRatings = GetRatings(premise.id, NodeRatingType.truth, userIDs);
		premiseTruthRatings = RatingListAfterRemovesAndAdds(premiseTruthRatings, ratingsBeingRemoved, ratingsBeingAdded, {nodeID: premise.id, ratingType: NodeRatingType.truth, userIDs});

		const usersWhoRatedPremise = new Set(premiseTruthRatings.map(a=>a.creator));
		for (const userID of usersWhoRatedArgAndPremises) {
			if (!usersWhoRatedPremise.has(userID)) {
				usersWhoRatedArgAndPremises.delete(userID);
			}
		}
	}

	const result = [] as NodeRating_MaybePseudo[];
	for (const userID of usersWhoRatedArgAndPremises) {
		const impactRating = GetArgumentImpactPseudoRating(argument, premises, userID, useAverageForMissing);
		if (impactRating) {
			result.push(impactRating);
		}
	}
	return result;
});

/* export function CalculateArgumentStrength(nodeChildren: NodeL1[]) {
	if (nodeChildren.Any(a=>a == null)) return 0; // must still be loading
	let impactPremise = nodeChildren.First(a=>a.impactPremise != null);
	let premises = nodeChildren.Except(impactPremise);
	if (premises.length == 0) return 0;

	let premiseProbabilities = premises.map(child=>GetRatingAverage(child._id, "probability", 0) / 100);
	let all = impactPremise.impactPremise.ifType == ImpactPremise_IfType.All;
	let combinedProbabilityOfPremises = all
		? premiseProbabilities.reduce((total, current)=>total * current, 1)
		: premiseProbabilities.Max(null, true);

	if (impactPremise.impactPremise.thenType == ImpactPremise_ThenType.StrengthenParent || impactPremise.impactPremise.thenType == ImpactPremise_ThenType.WeakenParent) {
		let averageAdjustment = GetRatingAverage(impactPremise._id, "adjustment", 50);
		let strengthForType = averageAdjustment.Distance(50) / 50;
		var result = combinedProbabilityOfPremises * strengthForType;
	} else {
		var result = combinedProbabilityOfPremises * (GetRatingAverage(impactPremise._id, "probability", 0) / 100);
	}
	return (result * 100).RoundTo(1);
} */

File: Packages/app-server/src/db/commands/_shared/rating_processor.rs
use std::collections::HashSet;

use rust_shared::{
	anyhow::{anyhow, bail, Error},
	db_constants::SYSTEM_POLICY_PUBLIC_UNGOVERNED_NAME,
	itertools::Itertools,
	utils::{db::uuid::new_uuid_v4_as_b64_id, time::time_since_epoch_ms_i64},
};

use crate::{
	db::{
		_shared::access_policy_target::AccessPolicyTarget,
		access_policies::get_system_access_policy,
		node_links::ClaimForm,
		node_ratings::{get_node_rating_by_user, get_node_rating_by_user_base, get_node_ratings, get_node_ratings_base, get_rating_average, get_rating_average_base, NodeRating},
		node_ratings_::_node_rating_type::NodeRatingType,
		nodes_::_node::{get_node_form, ArgumentType, Node},
	},
	utils::db::accessors::AccessorContext,
};

// sync:js
pub async fn get_argument_impact_pseudo_rating(ctx: &AccessorContext<'_>, argument: &Node, premises: &[Node], user_id: &str, use_average_for_missing: bool) -> Result<NodeRating, Error> {
	if premises.len() == 0 {
		return Err(anyhow!("No premises provided."));
	}

	let mut premise_probabilities = vec![];
	for premise in premises {
		let rating_value = match get_node_rating_by_user_base(ctx, premise.id.as_str(), NodeRatingType::truth, user_id).await {
			Ok(rating) => rating.value,
			Err(_) => match use_average_for_missing {
				true => get_rating_average_base(ctx, premise.id.as_str(), NodeRatingType::truth, None).await?,
				false => bail!("Premise node #{} has no truth rating by user {}.", premise.id.as_str(), user_id),
			},
		};

		let form = get_node_form(ctx, &premise.id, &argument.id).await?;
		let probability = match form {
			ClaimForm::negation => 1.0 - (rating_value / 100.0),
			_ => rating_value / 100.0,
		};
		premise_probabilities.push(probability);
	}

	let combined_truth_of_premises: f64 = match argument.argumentType {
		Some(ArgumentType::all) => premise_probabilities.iter().fold(1.0, |total, current| total * current),
		Some(ArgumentType::anyTwo) => {
			let strongest = premise_probabilities.iter().fold(0f64, |max, current| max.max(*current));
			let second_strongest = if premise_probabilities.len() > 1 {
				premise_probabilities.iter().fold(0f64, |max, current| if *current != strongest { max.max(*current) } else { max })
			} else {
				0f64
			};
			strongest * second_strongest
		},
		Some(ArgumentType::any) => premise_probabilities.iter().fold(0f64, |max, current| max.max(*current)),
		None => return Err(anyhow!("Argument node's `argumentType` field was set to null!")),
	};

	let relevance: f64 = match get_node_rating_by_user_base(ctx, argument.id.as_str(), NodeRatingType::relevance, user_id).await {
		Ok(rating) => rating.value,
		Err(_) => match use_average_for_missing {
			true => get_rating_average_base(ctx, &argument.id, NodeRatingType::relevance, None).await?,
			false => bail!("Argument node #{} has no relevance rating by user {}.", argument.id.as_str(), user_id),
		},
	};

	// let strengthForType = adjustment.Distance(50) / 50;
	let result: f64 = combined_truth_of_premises * (relevance / 100f64);

	let access_policy = get_system_access_policy(ctx, SYSTEM_POLICY_PUBLIC_UNGOVERNED_NAME).await?;
	Ok(NodeRating {
		id: new_uuid_v4_as_b64_id(),
		accessPolicy: access_policy.id.to_string(),
		node: argument.id.to_string(),
		r#type: NodeRatingType::impact,
		creator: user_id.to_owned(),
		createdAt: time_since_epoch_ms_i64(),
		value: float_to_percent(result),
		c_accessPolicyTargets: vec![], // auto-set by db
	})
}

fn float_to_percent(f: f64) -> f64 {
	let result = f * 100.0;
	result.round()
}

/*pub struct RatingFilter {
	pub node_id: String,
	pub rating_type: NodeRatingType,
	pub user_ids: Option<Vec<String>>,
}
pub fn rating_list_after_removes_and_adds(base_list: &[NodeRating], ratings_to_remove: Option<&Vec<String>>, ratings_to_add: Option<&Vec<NodeRating>>, ratings_to_add_filter: &RatingFilter) -> Vec<NodeRating> {
	let mut result = base_list.to_vec();
	if let Some(ratings_to_remove) = ratings_to_remove {
		result.retain(|a| !ratings_to_remove.iter().any(|b| b == a.id.as_str()));
	}
	if let Some(ratings_to_add) = ratings_to_add {
		let ratings_to_add_filtered = ratings_to_add.iter().filter(|a| {
			a.node == ratings_to_add_filter.node_id
			&& a.r#type == ratings_to_add_filter.rating_type
			&& (ratings_to_add_filter.user_ids.is_none() || ratings_to_add_filter.user_ids.unwrap().iter().any(|b| b == &a.creator))
		}).map(|a| a.clone()).collect_vec();
		result.extend(ratings_to_add_filtered);
	}
	result
}*/

// sync:js[loosely]
pub async fn get_argument_impact_pseudo_ratings(ctx: &AccessorContext<'_>, argument: &Node, premises: &Vec<Node>, user_ids: Option<&Vec<String>>, use_average_for_missing: bool) -> Result<Vec<NodeRating>, Error> {
	let mut result = vec![];

	let argument_relevance_ratings = get_node_ratings_base(ctx, &argument.id, Some(NodeRatingType::relevance), user_ids).await?;

	let mut users_who_rated_arg_and_premises = argument_relevance_ratings.iter().map(|a| a.creator.clone()).collect::<HashSet<String>>();
	for premise in premises {
		let premise_truth_ratings = get_node_ratings_base(ctx, &premise.id, Some(NodeRatingType::truth), user_ids).await?;

		let users_who_rated_premise = premise_truth_ratings.iter().map(|a| a.creator.clone()).collect::<HashSet<String>>();
		for user_id in users_who_rated_arg_and_premises.clone() {
			if !users_who_rated_premise.contains(&user_id) {
				users_who_rated_arg_and_premises.remove(&user_id);
			}
		}
	}

	for user_id in &users_who_rated_arg_and_premises {
		if let Ok(impact_rating) = get_argument_impact_pseudo_rating(ctx, &argument, &premises, user_id, use_average_for_missing).await {
			result.push(impact_rating);
		}
	}

	Ok(result)
}
